File: code/packages/python/cas-simplify/src/cas_simplify/canonical.py
"""Canonical-form pass over the symbolic IR.

The canonical pass performs *purely structural* normalization:

1. **Flatten** — ``Add(a, Add(b, c))`` becomes ``Add(a, b, c)`` (and
   the same for ``Mul``).
2. **Sort** — args of commutative heads are sorted by a stable
   deterministic key, so ``Add(c, a, b)`` becomes ``Add(a, b, c)``.
3. **Singleton drop** — ``Add(x)`` becomes ``x``, ``Mul(x)`` becomes
   ``x``. (The IR allows them; canonical removes them.)
4. **Empty container** — ``Add()`` becomes ``IRInteger(0)``, ``Mul()``
   becomes ``IRInteger(1)``.

No identity rules, no constant folding, no rewrite — that's
:func:`cas_simplify.simplify`'s job. Keeping the canonical pass pure
makes it cheap and idempotent: ``canonical(canonical(x)) == canonical(x)``.

The sort key is structural: a tuple ``(rank, repr)`` where ``rank``
groups by node type so the order is deterministic across runs without
depending on Python's ``id()``.
"""

from __future__ import annotations

from symbolic_ir import (
    IRApply,
    IRFloat,
    IRInteger,
    IRNode,
    IRRational,
    IRString,
    IRSymbol,
)

from cas_simplify.heads import is_commutative_flat

# ---------------------------------------------------------------------------
# Sort key
# ---------------------------------------------------------------------------


def _rank(node: IRNode) -> int:
    """Stable ordering by node type — integers first, then rationals,
    then floats, then symbols, then compound, then strings.

    Numeric literals come first so they cluster together, which is
    what users expect when they see ``2 + x + y``: the constant is on
    the left.
    """
    if isinstance(node, IRInteger):
        return 0
    if isinstance(node, IRRational):
        return 1
    if isinstance(node, IRFloat):
        return 2
    if isinstance(node, IRSymbol):
        return 3
    if isinstance(node, IRApply):
        return 4
    if isinstance(node, IRString):
        return 5
    return 6


def _sort_key(node: IRNode) -> tuple[int, str]:
    """Total order on IR nodes: rank, then string form."""
    return (_rank(node), repr(node))


# ---------------------------------------------------------------------------
# Public canonical pass
# ---------------------------------------------------------------------------


def canonical(node: IRNode) -> IRNode:
    """Recursively normalize ``node`` into canonical form.

    Idempotent: applying twice gives the same result as applying once.
    """
    if isinstance(node, IRApply):
        return _canonical_apply(node)
    return node


def _canonical_apply(node: IRApply) -> IRNode:
    # First, canonicalize the children.
    new_head = canonical(node.head)
    new_args = tuple(canonical(a) for a in node.args)

    if isinstance(new_head, IRSymbol) and is_commutative_flat(new_head.name):
        new_args = _flatten(new_head.name, new_args)
        new_args = tuple(sorted(new_args, key=_sort_key))

        if not new_args:
            # Add() → 0, Mul() → 1.
            return IRInteger(0 if new_head.name == "Add" else 1)
        if len(new_args) == 1:
            # Singleton drop: Add(x) → x, Mul(x) → x.
            return new_args[0]

    if new_head is node.head and new_args == node.args:
        return node
    return IRApply(new_head, new_args)


def _flatten(head_name: str, args: tuple[IRNode, ...]) -> tuple[IRNode, ...]:
    """Flatten any direct ``IRApply(head_name, ...)`` children into the args."""
    out: list[IRNode] = []
    for a in args:
        if (
            isinstance(a, IRApply)
            and isinstance(a.head, IRSymbol)
            and a.head.name == head_name
        ):
            out.extend(a.args)
        else:
            out.append(a)
    return tuple(out)

File: code/packages/rust/cas-simplify/src/canonical.rs
//! Canonical-form normalization pass.
//!
//! Four transformations, all purely structural:
//!
//! 1. **Flatten** — `Add(a, Add(b, c))` → `Add(a, b, c)` (same for `Mul`).
//! 2. **Sort** — args of commutative heads are sorted by a stable key,
//!    so `Add(c, a, b)` → `Add(a, b, c)`.
//! 3. **Singleton drop** — `Add(x)` → `x`, `Mul(x)` → `x`.
//! 4. **Empty container** — `Add()` → `0`, `Mul()` → `1`.
//!
//! Idempotent: `canonical(canonical(x)) == canonical(x)`.

use symbolic_ir::{IRApply, IRNode, ADD, MUL};

/// Heads that are commutative and should be flattened and sorted.
fn is_commutative_flat(name: &str) -> bool {
    name == ADD || name == MUL
}

/// The sort-key rank for an IR node.
///
/// ```text
/// Integer  → 0
/// Rational → 1
/// Float    → 2
/// Symbol   → 3
/// Apply    → 4
/// Str      → 5
/// ```
fn rank(node: &IRNode) -> u8 {
    match node {
        IRNode::Integer(_) => 0,
        IRNode::Rational(_, _) => 1,
        IRNode::Float(_) => 2,
        IRNode::Symbol(_) => 3,
        IRNode::Apply(_) => 4,
        IRNode::Str(_) => 5,
    }
}

/// Stable total order for IR nodes: (rank, display_string).
fn sort_key(node: &IRNode) -> (u8, String) {
    (rank(node), format!("{node:?}"))
}

/// Recursively normalize `node` into canonical form.
pub fn canonical(node: IRNode) -> IRNode {
    match node {
        IRNode::Apply(apply) => canonical_apply(*apply),
        other => other,
    }
}

fn canonical_apply(node: IRApply) -> IRNode {
    // Recursively canonicalize children.
    let new_head = canonical(node.head);
    let new_args: Vec<IRNode> = node.args.into_iter().map(canonical).collect();

    // Only flatten/sort for commutative associative heads.
    if let IRNode::Symbol(ref name) = new_head {
        if is_commutative_flat(name.as_str()) {
            let head_name = name.as_str();

            // 1. Flatten nested same-head children.
            let flat: Vec<IRNode> = new_args
                .into_iter()
                .flat_map(|a| flatten_one(a, head_name))
                .collect();

            // 2. Sort.
            let mut sorted = flat;
            sorted.sort_by(|a, b| sort_key(a).cmp(&sort_key(b)));

            // 3. Empty container.
            if sorted.is_empty() {
                return if head_name == ADD {
                    IRNode::Integer(0)
                } else {
                    IRNode::Integer(1)
                };
            }

            // 4. Singleton drop.
            if sorted.len() == 1 {
                return sorted.remove(0);
            }

            return IRNode::Apply(Box::new(IRApply {
                head: new_head,
                args: sorted,
            }));
        }
    }

    IRNode::Apply(Box::new(IRApply {
        head: new_head,
        args: new_args,
    }))
}

/// If `node` is `Apply(head_name, ...)`, yield its args; otherwise yield `node`.
fn flatten_one(node: IRNode, head_name: &str) -> Vec<IRNode> {
    if let IRNode::Apply(ref apply) = node {
        if let IRNode::Symbol(ref h) = apply.head {
            if h.as_str() == head_name {
                // Consume the node and return its (already-canonicalized) args.
                if let IRNode::Apply(boxed) = node {
                    return boxed.args;
                }
            }
        }
    }
    vec![node]
}
